kv/kvserver: TestPromoteNonVoterInAddVoter failed

kv/kvserver.TestPromoteNonVoterInAddVoter [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9606449?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9606449?buildTab=artifacts#/) on master @ [8e6cdbd8b3d12ba5d65db101a3abfeb2cb1bf470](https://github.com/cockroachdb/cockroach/commits/8e6cdbd8b3d12ba5d65db101a3abfeb2cb1bf470):


```
=== RUN   TestPromoteNonVoterInAddVoter
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/d486233d291fa43f53db749aea077e75/logTestPromoteNonVoterInAddVoter3018638515
    test_log_scope.go:79: use -show-logs to present logs inline
    replicate_queue_test.go:2349: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:2349
        	Error:      	Expected to promote non-voter, instead added voter
        	Test:       	TestPromoteNonVoterInAddVoter
        	Messages:   	Added voter store ID: 6
        	            	Add voter events: [{r63:/{Table/104-Max} [(n3,s3):1, (n2,s2):2, (n1,s1):7, (n6,s6):8NON_VOTER, (n4,s4):9, next=10, gen=32] <nil> <nil> (n4,s4):9 <nil> range under-replicated {"Target":"s4, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:62"}} {r63:/{Table/104-Max} [(n3,s3):1, (n2,s2):2, (n1,s1):7, (n6,s6):8, (n4,s4):9, next=10, gen=33] <nil> <nil> (n6,s6):8 <nil> range under-replicated {"Target":"s6, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:62"}} {r63:/{Table/104-Max} [(n5,s5):10, (n2,s2):2, (n1,s1):7, (n6,s6):11LEARNER, next=12, gen=44] <nil> <nil> (n6,s6):11LEARNER <nil> range under-replicated {"Target":"s6, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:61"}} {r63:/{Table/104-Max} [(n5,s5):10, (n2,s2):2, (n1,s1):7, (n6,s6):11, next=12, gen=45] <nil> <nil> (n6,s6):11 <nil> range under-replicated {"Target":"s6, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:61"}} {r63:/{Table/104-Max} [(n5,s5):10, (n2,s2):2, (n1,s1):7, (n6,s6):11, (n4,s4):12LEARNER, next=13, gen=46] <nil> <nil> (n4,s4):12LEARNER <nil> range under-replicated {"Target":"s4, necessary:true, diversity:0.88, converges:0, balance:0, rangeCount:61"}} {r63:/{Table/104-Max} [(n5,s5):10, (n2,s2):2, (n1,s1):7, (n6,s6):11, (n4,s4):12, next=13, gen=47] <nil> <nil> (n4,s4):12 <nil> range under-replicated {"Target":"s4, necessary:true, diversity:0.88, converges:0, balance:0, rangeCount:61"}}]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d486233d291fa43f53db749aea077e75/logTestPromoteNonVoterInAddVoter3018638515
--- FAIL: TestPromoteNonVoterInAddVoter (30.90s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPromoteNonVoterInAddVoter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26986

kv/kvserver.TestPromoteNonVoterInAddVoter [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9841895?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9841895?buildTab=artifacts#/) on master @ [4bbe2faae22d21a650fbdbc280e0f3b2ad81aaf6](https://github.com/cockroachdb/cockroach/commits/4bbe2faae22d21a650fbdbc280e0f3b2ad81aaf6):


```
=== RUN   TestPromoteNonVoterInAddVoter
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/b5b16660ca668d68f7803e1ac963747e/logTestPromoteNonVoterInAddVoter3656602792
    test_log_scope.go:79: use -show-logs to present logs inline
    replicate_queue_test.go:2350: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:2350
        	Error:      	Expected to promote non-voter, instead added voter
        	Test:       	TestPromoteNonVoterInAddVoter
        	Messages:   	Added voter store ID: 5
        	            	Add voter events: [{r64:/{Table/104-Max} [(n1,s1):1, (n2,s2):4, (n3,s3):7, (n7,s7):8NON_VOTER, (n4,s4):9, next=10, gen=28] <nil> <nil> (n4,s4):9 <nil> range under-replicated {"Target":"s4, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:63"}} {r64:/{Table/104-Max} [(n1,s1):1, (n2,s2):4, (n3,s3):7, (n7,s7):8, (n4,s4):9, next=10, gen=29] <nil> <nil> (n7,s7):8 <nil> range under-replicated {"Target":"s7, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:63"}} {r64:/{Table/104-Max} [(n4,s4):9, (n2,s2):4, (n3,s3):7, (n7,s7):8, (n5,s5):11LEARNER, next=12, gen=37] <nil> <nil> (n5,s5):11LEARNER <nil> range under-replicated {"Target":"s5, necessary:true, diversity:0.88, converges:0, balance:0, rangeCount:62"}} {r64:/{Table/104-Max} [(n4,s4):9, (n2,s2):4, (n3,s3):7, (n7,s7):8, (n5,s5):11, next=12, gen=38] <nil> <nil> (n5,s5):11 <nil> range under-replicated {"Target":"s5, necessary:true, diversity:0.88, converges:0, balance:0, rangeCount:62"}}]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/b5b16660ca668d68f7803e1ac963747e/logTestPromoteNonVoterInAddVoter3656602792
--- FAIL: TestPromoteNonVoterInAddVoter (32.09s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPromoteNonVoterInAddVoter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestPromoteNonVoterInAddVoter [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9981861?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9981861?buildTab=artifacts#/) on master @ [adad57859a5281e2846082761c8232890b3d2a68](https://github.com/cockroachdb/cockroach/commits/adad57859a5281e2846082761c8232890b3d2a68):


```
=== RUN   TestPromoteNonVoterInAddVoter
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/8af8a6c853877a083ec84004541c621b/logTestPromoteNonVoterInAddVoter1191499411
    test_log_scope.go:79: use -show-logs to present logs inline
    replicate_queue_test.go:2350: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:2350
        	Error:      	Expected to promote non-voter, instead added voter
        	Test:       	TestPromoteNonVoterInAddVoter
        	Messages:   	Added voter store ID: 7
        	            	Add voter events: [{r64:/{Table/104-Max} [(n3,s3):1, (n2,s2):4, (n1,s1):7, (n5,s5):8, (n7,s7):9NON_VOTER, next=10, gen=32] <nil> <nil> (n5,s5):8 <nil> range under-replicated {"Target":"s5, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:63"}} {r64:/{Table/104-Max} [(n3,s3):1, (n2,s2):4, (n1,s1):7, (n5,s5):8, (n7,s7):9, next=10, gen=33] <nil> <nil> (n7,s7):9 <nil> range under-replicated {"Target":"s7, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:63"}} {r64:/{Table/104-Max} [(n4,s4):10, (n2,s2):4, (n1,s1):7, (n5,s5):8, (n7,s7):11LEARNER, next=12, gen=41] <nil> <nil> (n7,s7):11LEARNER <nil> range under-replicated {"Target":"s7, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:62"}} {r64:/{Table/104-Max} [(n4,s4):10, (n2,s2):4, (n1,s1):7, (n5,s5):8, (n7,s7):11, next=12, gen=42] <nil> <nil> (n7,s7):11 <nil> range under-replicated {"Target":"s7, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:62"}}]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/8af8a6c853877a083ec84004541c621b/logTestPromoteNonVoterInAddVoter1191499411
--- FAIL: TestPromoteNonVoterInAddVoter (30.29s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPromoteNonVoterInAddVoter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestPromoteNonVoterInAddVoter [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10004273?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10004273?buildTab=artifacts#/) on master @ [0d890b56341ca40b5ffb37c51d6b63bc322e80f4](https://github.com/cockroachdb/cockroach/commits/0d890b56341ca40b5ffb37c51d6b63bc322e80f4):


```
=== RUN   TestPromoteNonVoterInAddVoter
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/ac3e4004160555c7c575e1264d1b8f5f/logTestPromoteNonVoterInAddVoter1212694418
    test_log_scope.go:79: use -show-logs to present logs inline
    replicate_queue_test.go:2357: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:2357
        	Error:      	Expected to promote non-voter, instead added voter
        	Test:       	TestPromoteNonVoterInAddVoter
        	Messages:   	Added voter store ID: 5
        	            	Add voter events: [{r64:/{Table/104-Max} [(n1,s1):1, (n2,s2):4, (n3,s3):7, (n6,s6):8, (n5,s5):9NON_VOTER, next=10, gen=28] <nil> <nil> (n6,s6):8 <nil> range under-replicated {"Target":"s6, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:63"}} {r64:/{Table/104-Max} [(n1,s1):1, (n2,s2):4, (n3,s3):7, (n6,s6):8, (n5,s5):9, next=10, gen=29] <nil> <nil> (n5,s5):9 <nil> range under-replicated {"Target":"s5, necessary:true, diversity:1.00, converges:0, balance:0, rangeCount:63"}} {r64:/{Table/104-Max} [(n4,s4):10, (n2,s2):4, (n3,s3):7, (n6,s6):8, (n5,s5):11LEARNER, next=12, gen=37] <nil> <nil> (n5,s5):11LEARNER <nil> range under-replicated {"Target":"s5, necessary:true, diversity:0.88, converges:0, balance:0, rangeCount:62"}} {r64:/{Table/104-Max} [(n4,s4):10, (n2,s2):4, (n3,s3):7, (n6,s6):8, (n5,s5):11, next=12, gen=38] <nil> <nil> (n5,s5):11 <nil> range under-replicated {"Target":"s5, necessary:true, diversity:0.88, converges:0, balance:0, rangeCount:62"}}]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/ac3e4004160555c7c575e1264d1b8f5f/logTestPromoteNonVoterInAddVoter1212694418
--- FAIL: TestPromoteNonVoterInAddVoter (31.10s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPromoteNonVoterInAddVoter.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/allocator/plan/replicate.go
+++ b/pkg/kv/kvserver/allocator/plan/replicate.go
@@ -144,6 +144,10 @@ func (rp ReplicaPlanner) ShouldPlanChange(
 	canTransferLeaseFrom CanTransferLeaseFrom,
 ) (shouldPlanChange bool, priority float64) {
 	desc, conf := repl.DescAndSpanConfig()
+
+	log.KvDistribution.VEventf(ctx, 6,
+		"computing range action desc=%s config=%s",
+		desc, conf.String())
 	action, priority := rp.allocator.ComputeAction(ctx, rp.storePool, conf, desc)
 
 	if action == allocatorimpl.AllocatorNoop {
@@ -249,6 +253,9 @@ func (rp ReplicaPlanner) PlanOneChange(
 	// successfully execute a decision that was based on the state of a stale
 	// range descriptor.
 	desc, conf := repl.DescAndSpanConfig()
+	log.KvDistribution.VEventf(ctx, 6,
+		"planning range change desc=%s config=%s",
+		desc, conf.String())
 
 	voterReplicas, nonVoterReplicas,
 		liveVoterReplicas, deadVoterReplicas,

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -2194,13 +2194,13 @@ func iterateOverAllStores(
 // non-voters when switching from ZONE to REGION survival i.e.
 //
 // ZONE survival configuration:
-// Region 1: Voter, Voter, Voter
-// Region 2: Non-Voter
-// Region 3: Non-Voter
+// Region 1: 3 of [n1 (voter) n2 (voter) n3 (voter)]
+// Region 2: 1 of [n4 (non-voter) n5 (non-voter)]
+// Region 3: 1 of [n6 (non-voter) n7 (non-voter)]
 // to REGION survival configuration:
-// Region 1: Voter, Voter
-// Region 2: Voter, Voter
-// Region 3: Voter
+// Region 1: 2 of [n1 (voter) n2 (voter) n3 (voter)]
+// Region 2: 2 of [n4 (voter) n5 (voter)]
+// Region 3: 1 of [n6 (voter) n7 (voter)]
 //
 // Here we have 7 stores: 3 in Region 1, 2 in Region 2, and 2 in Region 3.
 //
@@ -2345,8 +2345,44 @@ func TestPromoteNonVoterInAddVoter(t *testing.T) {
 	var rangeID roachpb.RangeID
 	err = db.QueryRow("SELECT range_id FROM [SHOW RANGES FROM TABLE t] LIMIT 1").Scan(&rangeID)
 	require.NoError(t, err)
-	addVoterEvents, err := filterRangeLog(tc.Conns[0], rangeID, kvserverpb.RangeLogEventType_add_voter, kvserverpb.ReasonRangeUnderReplicated)
+	addVoterEvents, err := filterRangeLog(tc.Conns[0],
+		rangeID, kvserverpb.RangeLogEventType_add_voter, kvserverpb.ReasonRangeUnderReplicated)
 	require.NoError(t, err)
+
+	// If there are more than 2 add voter events, it implies that we ran into an
+	// issue where we likely down-replicated from the desired end state of
+	// voters=5, then noticed this and subsequently up-replicated to recover back
+	// to voters=5. This can happen due to ill timed span config updates e.g.
+	//
+	//   Have the correct number of voters (5), however over-satisfied on voters
+	//   in Region 1 (3/2) and undersatisfied in Region 2 (1/2).
+	//
+	//   voters = [s1, s2, s3*, s5, s6] non = []
+	//
+	//   A rebalance occurs at t2 towards s4 from s3 to correct the over/under
+	//   satisfaction. The lease also transfers with the rebalance. Now at Region
+	//   1 (2/2), Region 2 (2/2) and  Region 3 (1/1).
+	//
+	//   voters = [s1, s2, s4*, s5, s6] non = []
+	//
+	//   However, the new leaseholder store s4 has not received the region
+	//   survival config changes and still sees the old zone survival config. s4
+	//   proceeds to start removing voters as there only need to be 3.
+	//
+	//   voters = [s1, s2, s4*, s6] non = [] (note we don't demote here)
+	//
+	//   s4 receives the update that s1 made their initial changes on (voters=5)
+	//   and proceeds to add a voter back to 5.
+	//
+	//   voters = [s1, s2, s4*, s5, s6] non = []
+	//
+	// This is unfortunate but the impact should be limited, so long as the new
+	// leaseholder receives the span config update within a short period of time.
+	// See #101519. If there are more than 2 add voter events, check only the
+	// first 2.
+	if len(addVoterEvents) > 2 {
+		addVoterEvents = addVoterEvents[:2]
+	}
 
 	// Check if an add voter event has an added replica of type LEARNER, and if
 	// it does, it shows that we are adding a new voter rather than promoting an

